Windows PowerShell Compatibility Utilities (#1)
anmenaga/PowerShell-RFC: 1-Draft/RFCNNN-WindowsPS-Compatibility-Utilities.md

---
RFC:
Author: Andrew Menagarishvili
Status: Draft
SupercededBy:
Version: 0.1
Area: Language
Comments Due: 10/10/2019
Plan to implement: Yes
---

> # Windows PowerShell Compatibility Utilities

Reviewer: I think the title should be more user focused like: Importing Windows PowerShell Modules in PowerShell 7

Author: renamed.

> One of the big factors preventing existing Windows PowerShell users from moving to PowerShell Core has been cmdlet coverage. Specifically being able to use existing Windows PowerShell modules in PowerShell Core. Ease of UX is important here - working with 'WindowsPS-only' modules from PS Core should not be any different than working with 'PS Core - compartible' modules. Also this should not require any setup from the user.

Reviewer: ```suggestion
One of the big factors preventing existing Windows PowerShell users from moving to PowerShell Core has been cmdlet coverage. Specifically being able to use existing Windows PowerShell modules in PowerShell Core. Ease of UX is important here - working with 'WindowsPS-only' modules from PS Core should not be any different than working with 'PS Core - compatible' modules. Also this should not require any setup from the user.
```

> One of the big factors preventing existing Windows PowerShell users from moving to PowerShell Core has been cmdlet coverage. Specifically being able to use existing Windows PowerShell modules in PowerShell Core. Ease of UX is important here - working with 'WindowsPS-only' modules from PS Core should not be any different than working with 'PS Core - compartible' modules. Also this should not require any setup from the user.

Reviewer 2: It might be good to mention that the non-compatibility is due to the different dotNet runtimes used between Windows PowerShell and PowerShell Core.

Author: Added.

## Motivation

    As a PowerShell user,
    I can use modules written for Windows PowerShell in PowerShell Core,
    so that I could migrate from Windows PowerShell to PowerShell Core.

## User Experience

Example that shows using commands from 'WindowsPS-only' module in PS Core:
```PowerShell
PS C:\> $PSVersionTable.PSEdition
Core

> PS C:\> Import-Module DesktopOnlyModule -AsWindowsCompatibility

Reviewer: We shouldn't require a switch, but we should probably show a warning message so they understand the objects are deserialized

Author: removed the switch;
added verbose notification on import about deserialized objects. I feel like a verbose msg is a better fit than a warning msg for this use case.

Initial proposal was that absense of `Core` value in `CompatiblePSEditions` during module import would signal to PS Core that it should process the module as described in this RFC, however PS Core still allows to load non-`Core` modules without any warnings if they are not located under System32 module path. This means that `CompatiblePSEditions` property can not be used for deciding whether to load a module as WinPS module or not.
Current plan is to add a new `-AsWindowsCompatibility` parameter for `Import-Module`.

When Import-Module in PS Core detects that user is attempting to load a module with `-AsWindowsCompatibility` parameter:
  1. it creates a hidden `WindowsPS Compatibility` Windows PS process (unless one already exists)
  2. creates PS Remoting connection to its IPC pipe
  3. loads the module in remote Windows PS process
  4. generates local proxy module/commands using IPC remoting connection and `Import-PSSession` cmdlet
  5. when 'WindowsPS-only' module unload request is detected - unload module in remote process, close PS remoting channel, close remote process

### PS Remoting Transport

> IPC / Named Pipe is to be used for connections to remote PowerShell process. Today a named pipe listener is created with each Windows PowerShell process. IPC transport has good performance and secured endpoints, however (unlike, for example, SSH transport) it is not currently supported by New-PSSession.

Reviewer: I think you should call out here how this is different from the WindowsCompatibility module.  Specifically it doesn't rely on WinRM remoting.

Author: Looks like this is covered in 'Alternate Proposals and Considerations'.

> So a new IPC-specific parameter set will be added to New-PSSession:  `New-PSSession [-Name <String>] -ProcessId <Int32>`

Reviewer 2: We may want to add (maybe later) an optional AppDomainName parameter, to allow connecting to Windows PowerShell in a non-default app domain.

Author: It's probably good to have it; Added AppDomainName parameter.

Here is the pseudo-code that shows the essence of proposed implementation of a new parameter set in New-PSSession:
```csharp

> string WindowsPS_AppDomainName = GetPSHostProcessInfoCommand.GetAppDomainNamesFromProcessId(WindowsPS_ProcessId);

Reviewer 2: Technical detail:  We won't need to pass in `WindowsPS_AppDomainName`, because we will be creating the target WindowsPS process and PowerShell will run in the default appdomain, which NamedPipeConnection connects to by default.

Author: Ok; I removed obtaining 'WindowsPS_AppDomainName' from pseudo-code, but left it in calling NamedPipeConnectionInfo constructor as this pseudo-code is a about of 'New-PSSession -ProcessId <Int32> [-Name <String>] [-AppDomainName <String>]` implementation in general, rather than implementation of WinPSCompat scenario.

NamedPipeConnectionInfo connectionInfo = new NamedPipeConnectionInfo(WindowsPS_ProcessId, WindowsPS_AppDomainName);
TypeTable typeTable = TypeTable.LoadDefaultTypeFiles();
RemoteRunspace remoteRunspace = RunspaceFactory.CreateRunspace(connectionInfo, this.Host, typeTable) as RemoteRunspace;
remoteRunspace.Name = NamedPipeRunspaceName;

> remoteRunspace.ShouldCloseOnPop = true

Reviewer 2: Nitpick, We don't technically need to set this since we are not pushing the remote runspace, but it doesn't hurt anything.

Author: Ok; I wan't sure if this was important.  Removed.

try
{
    remoteRunspace.Open();
}
catch (RuntimeException e)
{
    // handle connection errors
}

> PSSession remotePSSession = new PSSession(remoteRunspace);

Reviewer 2: Might want to mention that `Import-PSSession` is then run on `remotePSSession` to generate proxy cmdlets.

Author: Added.

this.RunspaceRepository.Add(remotePSSession);
WriteObject(remotePSSession);
```

### Lifetime of remote Windows PowerShell process and module

> Overall RFC goal is to have farmiliar 'local module' user experience even though actuall operations are working with a module in a remote process. This drives following:

Reviewer: ```suggestion
Overall RFC goal is to have familiar 'local module' user experience even though actual operations are working with a module in a remote process. This drives following:
```

1. One remote Windows PS process corresponds to one local PS Core process. E.g. if a same user creates 2 local PS Core processes and loads 'WindowsPS-only' module in each one, this will result in creating 2 Windows PS processes and loading the actuall module in each one of them.

> 2. Lifetime of the module on remote side (Windows PS) should be the same as on local side (PS Core). This is important because some modules save their global state in-between command calls.

Reviewer: I wonder if using 'remote' will cause confusion even thought it goes through the PS remoting layer.  Would prefer just to call it local WinPS process.

Author: I feel like specifics of this WinPS process should be somehow marked, so I changed 'remote' to 'compatibility'.

   * Local Load-Module for 'WindowsPS-only' module will have same behaviour/effect in remote Windows PS process as it would have been done locally on an compartible module. In addition, this will create `WindowsPS Compatibility` Windows PS process if one does not already exist.

>    * Local Remove-Module for 'WindowsPS-only' module will have same behaviour/effect in remote Windows PS process as it would have been done locally on an compartible module. In addition, this will remove `WindowsPS Compatibility` Windows PS process if it is no longer needed.

Reviewer 2: I wonder if it would be better to create the WindowsPS process once and leave it until PowerShell Core exits?  It would save creating/exiting processes if the user adds/removes/adds/... modules.

In any case, I imagine the common scenario is to load the WindowsPS module and leave it for the duration of the PowerShell instance life.  The tricky part is to not leave the WindowsPS process dangling after PowerShell core exits (see 4. comment below).

Author: That is a good question; I don't have a strong opinion on either approach.
Feel free to post this comment in the official RFC PR after I create it. And we'll look at other people's input on this.

3. When removing a module, there is an `OnRemove` event on the module that will execute. This event allows Compat proxy module to react to being removed and perform cleanup of remote process if necessary.

> 4. PS process exit does Not perform gracefull cleanup of modules, so it needs to be handled separately. There is an event that allows to react to the closing of the PowerShell process (to do remote Compat process cleanup): `Register-EngineEvent -SourceIdentifier ([System.Management.Automation.PsEngineEvent]::Exiting) -Action $OnRemoveScript`

Reviewer: ```suggestion
4. PS process exit does Not perform graceful cleanup of modules, so it needs to be handled separately. There is an event that allows to react to the closing of the PowerShell process (to do remote Compat process cleanup): `Register-EngineEvent -SourceIdentifier ([System.Management.Automation.PsEngineEvent]::Exiting) -Action $OnRemoveScript`
```

> 4. PS process exit does Not perform gracefull cleanup of modules, so it needs to be handled separately. There is an event that allows to react to the closing of the PowerShell process (to do remote Compat process cleanup): `Register-EngineEvent -SourceIdentifier ([System.Management.Automation.PsEngineEvent]::Exiting) -Action $OnRemoveScript`

Reviewer 2: My main concern is a dangling WindowsPS process after PowerShell Core exits.  I don't think there is an OS way to guarantee a child process exits after the parent process exits.  That was one nice thing about WinRM service, as it could notify notify parent if child exited and terminate the child process if the parent process abruptly exited.

But I don't think we need a remoting event, and instead can use the Process.Exited event.  The child could monitor the parent and if the parent (PowerShell Core) exits unexpectedly then the child process can exit.  Or we could have some other clean up mechanism to ensure child processes are terminated at some point.

Author: Good point. Monitoring ParentProcess.Exited from child should work.
